listStakePools - Include metadata in API responses

rvl opened the pull request:
Relates to #1065.

# Overview

- Modified Api.Server.listPools and StakePoolLayer listStakePools to include metadata.
- Made it possible to use an environment variable for the registry url.
- Handles the case where different stake pool owners register different metadata by logging and then ignoring all inconsistent data.

# Comments

- This is a draft.

rvl committed:
Add metadata to listStakePools

rvl committed:
Cardano.Pool.Metadata: Add optional env var for the registry URL

rvl committed:
Cardano.Pool.Metrics: Add more structured logging

rvl committed:
Make associateMetadata into a pure function

KtorZ commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -261,6 +266,13 @@ findArchiveFile repoPath = find isRepoFile . Map.keys <$> getEntries
+getRegistryZipUrl :: IO String
+getRegistryZipUrl = fromMaybe cardanoFoundationRegistryZip <$> lookupEnv var
+  where
+    var = "CARDANO_WALLET_STAKE_POOL_REGISTRY_URL"
```
:+1:

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -211,12 +231,20 @@ data ErrListStakePools
 newStakePoolLayer db@DBLayer{..} nl tr = StakePoolLayer
     { listStakePools = do
+        lift $ logTrace tr MsgListStakePoolsBegin
+        stakePools <- rankKnownPools
```
**[remark]** 

You have used to forbidden word (i.e. `rank`) :scream: 
It makes things clearer though :+1:

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -238,8 +266,25 @@ newStakePoolLayer db@DBLayer{..} nl tr = StakePoolLayer
+        getStakePoolMetadata tr' url owners' >>= \case
+            Left _ -> do
+                logTrace tr MsgMetadataUnavailable
+                pure $ replicate (length poolIds) Nothing
```
**[remark]**

Isn't `trace` maybe too low for this kind of error? We are not really expecting this to be unavailable, are we? 
I'd rather go for `Warning` or `Notice` for this one. Though `Warning` is maybe too strong because there is really nothing we can do about it.

rvl replied:
The log level is set in `instance DefineSeverity StakePoolLayerMsg`. I had made it Warning but actually now that you mention it, Notice is more suitable.

KtorZ replied:
Ho! Yes. I missed that sorry, got confused with `logTrace` :+1:

KtorZ committed:
expose ENV var name for overriding the registry URL

KtorZ committed:
add integration test metadata registry for integration pools

KtorZ committed:
start mock registry server alongside the integration server in the integration test setup

rvl committed:
Add unit tests for associateMetadata

KtorZ committed:
make sure to account for registrations in the genesis block

KtorZ commented on lib/core/test/unit/Cardano/Pool/MetricsSpec.hs:
```diff
@@ -485,3 +497,69 @@ arbitraryChunks xs = do
+                (Map.fromList [(pid, [owner])])
+                [(owner, Just md)]
+
+        res `shouldBe` [(MsgMetadataUsing pid owner md, Just md)]
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Pool/MetricsSpec.hs:
```diff
@@ -485,3 +497,69 @@ arbitraryChunks xs = do
+                (Map.fromList [(PoolId "1", [PoolOwner "a"])])
+                [(PoolOwner "a", Nothing)]
+
+        res `shouldBe` [(MsgMetadataMissing (PoolId "1"), Nothing)]
```
:+1:

KtorZ commented on lib/core/test/unit/Cardano/Pool/MetricsSpec.hs:
```diff
@@ -485,3 +497,69 @@ arbitraryChunks xs = do
+                md `shouldSatisfy` sameStakePoolMetadata mdb
+            _ -> fail $ "wrong log message " ++ show msg
+
+        fmap (sameStakePoolMetadata mda) res `shouldBe` Just True
```
:+1:

KtorZ committed:
factor out jcli logic and refactor Jormungandr faucet to use this

KtorZ committed:
fixture for easily publishing new pool certificate on chain in isolation with other tests

KtorZ committed:
add integration tests checking that newly registered stake pools are treated correctly

KtorZ committed:
do not use a Map for pool owners! We need to preserve the list order for zipping correctly!

KtorZ committed:
re-generate nix machinery

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -170,6 +183,30 @@ spec = do
             expectResponseCode @IO HTTP.status405 r
             expectErrorMessage errMsg405 r
 
+    it "STAKE_POOLS_LIST_04 - Discovers new pools when they are registered" $ \ctx -> do
```
@rvl I added a rather extensive integration test. The setup behind `registerStakePool` is a bit complex because we have to do everything with `jcli` and use accounts to pay for the registration certificate! I got to the end of it and.. nicely caught a small bug in the "zipping logic" from the `listStakePools` endpoint (cf my last commit). 
Please have another round of review / fixes and let's get this merged! 

Some outstanding points remaining after that:

- Maybe do some caching? I expect this endpoint to be quite heavily hammered by clients like Daedalus. The registry won't change very often. Renewing a cache every hour or so should also mitigate the concern. 
- Currently the handler for `joinStakePool` relies on `listStakePool` to get a list of known stake pool... that's kinda lame because `listStakePool` is a very heavy operation and, all stake pools ids can be known by simply making a database call. So we should fix that.
- https://github.com/input-output-hk/cardano-wallet/issues/1158

rvl committed:
Update stake pool registry test data for unit tests

https://github.com/input-output-hk/testnet-stake-pool-registry/archive/master.zip

Current revision: c6d196f630b28869bffb9e3af453a9c65be2fa7c

rvl committed:
Add Tests.Utils.StaticServer.withStaticServer

rvl committed:
Fix metadata in STAKE_POOLS_LIST_01

rvl committed:
Update MetadataSpec with new test registry data
